src/server/.tests/mocha/routes/controllers/email.spec.js

```diff
@@ -0,0 +1,192 @@
+        it('Should send an e-mail with the specified details.', () => {
+            sandbox.stub(nodemailer, 'createTransport').returns(DUMMY_TRANSPORT)
+            const transport = nodemailer.createTransport()
+            emailController.sendMail(transport)(DUMMY_OPTIONS, (err) => {
+                assert.strictEqual(err, null)
+            })
+        })
+        it('Should throw an error when no options are provided.', () => {
+            assert.throws(() => emailController.sendMail(DUMMY_TRANSPORT)(NO_EMAIL), Error, EXPECTED_ERROR_MESSAGE)
+        })
+    })
+    describe('Save Record to DB', () => {
+        let suite = {}
+        const verifyMock = function(){
+            return {
+                save: sinon.stub().returns(new Promise(res => res()))
+            }
+        }
+        beforeEach(() => {
+            suite.saveRecord = emailController.saveRecordToDB(verifyMock)
+        })
+        afterEach(() => {
+            suite = {}
+        })
+        it('Should return true if all inputs are valid.', () => {
+            const ACTUAL_RESULT = suite.saveRecord(DUMMY_USER, DUMMY_TOKEN)
+            const EXPECTED_RESULT = true
+
+            assert(ACTUAL_RESULT, EXPECTED_RESULT)
+        })
+        it('Should throw an error when no details are provided.', () => {
+            assert.throws(() => suite.saveRecord(INVALID_USER, INVALID_TOKEN))
+        })
+    })
+    describe('Send Verification Mail', () => {
+        let suite = {}
+        const DUMMY_TRANSPORT = {
+            sendMail: sinon.spy()
+        }
+        
+        beforeEach(() => {
+            suite.send = emailController.sendVerificationMail(DUMMY_TRANSPORT)
+        })
+        afterEach(() => {
+            suite = {}
+        })
+        it('Should throw an error when invalid details are provided', () => {
+            assert.throws(() => suite.send(DUMMY_EMAIL, DUMMY_HOST, INVALID_TOKEN))
+        })
+        it('Should send an email when details are correct', () => {
+            suite.send(DUMMY_EMAIL, DUMMY_HOST, DUMMY_TOKEN)
+            sinon.assert.calledOnce(DUMMY_TRANSPORT.sendMail)
+        })
+    })
+    describe('Email Verification', () => {
+        let suite = {}
+
+        beforeEach(() => {
+            const verifyMock = {
+                find: function(token, callback){
+                    return callback(null, {username: DUMMY_USER})
+                },
+                remove: function(token, callback){
+                    return callback(null)
+                }
+            }
+
+            const userMock = {
+                findOneAndUpdate: function(username, action, callback){
+                    return callback(null)
+                }
+            }
+            suite.verify = emailController.verifyUser(verifyMock, userMock)
+        })
+        afterEach(() => {
+            suite = {}
+        })
+
+        it('Should return a BAD_REQUEST message if no token is provided.', () => {
```

Comment:
When does verify runs? Isn't it too deep inside server logic to return HTTP error?
Shouldn't it send some different error that's later being mapped to BAD_REQUEST?

Reply:
This is correct, I'll remove them and make them throw an error so that it triggers the register controller's catch method, which will return a bad request. 

Some food for thought, but I believe we should make these bad requests more meaningful as they're being used to hide a lot of different kinds of possible errors at this stage.